CMakeLists.txt

```diff
@@ -3,6 +3,17 @@ project(mlpack C CXX)
 
 include(CMake/cotire.cmake)
 
+# Validate md5 hash give md5file url and the file directory.
+# If hash check fails set HASH_CHECK_FAIL to non-zero and remove the files.
+macro (check_hash MD5_URL DIR HASH_CHECK_FAIL)
+  file(DOWNLOAD MD5_URL "${DIR}/hash.md5")
+  execute_process(COMMAND ${CMAKE_COMMAND} -E md5sum -c "${DIR}/hash.md5"
+      RESULT_VARIABLE HASH_CHECK_FAIL)
```

Comment:
Hmm, so this doesn't actually run the `md5sum` program; see the output from Travis:

```
-c: No such file or directory

d41d8cd98f00b204e9800998ecf8427e  /home/travis/build/mlpack/mlpack/build/deps/stb/hash.md5

CMake Warning at CMakeLists.txt:378 (message):

  Could not download stb! Error code : "No error"! Error log:

CMake Warning at CMakeLists.txt:380 (message):

  stb/stb_image.h is not installed.  Image utilities will not be available!

```

If you use `${CMAKE_COMMAND} -E md5sum`, the documentation suggests that this only creates the md5sum.  So, I think it'll be a little bit uglier, but maybe the right thing to do here is use CMake's `file()` and `string()` functions along with `${CMAKE_COMMAND} -E md5sum` in order to generate the md5sum, and then compare with the md5sums downloaded in `hash.md5`.

That'll probably be a good bit of CMake code, so maybe it's worth splitting it into its own file in `CMake/` to try and reduce the amount of stuff in this file. :+1:

Reply:
I moved the macro to CMake/CheckHash.cmake
[md5sum](https://linux.die.net/man/1/md5sum) can also check the md5 files created by it.